File: ios/PiAgentCore/Sources/PiAgentCore/AgentTypes.swift
import Foundation
import PiAI

// MARK: - Tool Protocol

public protocol Tool: Sendable {
    var name: String { get }
    var description: String { get }
    var parametersSchema: JSONValue { get }
    func execute(input: JSONValue) async throws -> AgentToolResult
}

// MARK: - Tool Result

public struct AgentToolResult: Sendable {
    public let toolCallId: String
    public let output: String
    public let details: ToolResultDetails?
    public let isError: Bool

    public init(toolCallId: String = "", output: String, details: ToolResultDetails? = nil, isError: Bool = false) {
        self.toolCallId = toolCallId
        self.output = output
        self.details = details
        self.isError = isError
    }
}

// MARK: - Tool Result Details

public enum ToolResultDetails: Sendable {
    case file(path: String, content: String, language: String?)
    case diff(path: String, hunks: [DiffHunk])
    case table(columns: [String], rows: [[String]])
    case error(message: String, code: String?)
}

// MARK: - Diff Types

public struct DiffHunk: Sendable {
    public let startLineOld: Int
    public let countOld: Int
    public let startLineNew: Int
    public let countNew: Int
    public let lines: [DiffLine]

    public init(startLineOld: Int, countOld: Int, startLineNew: Int, countNew: Int, lines: [DiffLine]) {
        self.startLineOld = startLineOld
        self.countOld = countOld
        self.startLineNew = startLineNew
        self.countNew = countNew
        self.lines = lines
    }
}

public struct DiffLine: Sendable {
    public let type: DiffLineType
    public let content: String

    public init(type: DiffLineType, content: String) {
        self.type = type
        self.content = content
    }
}

public enum DiffLineType: Sendable {
    case context
    case add
    case remove
}

// MARK: - Agent Events

public enum AgentEvent: Sendable {
    case streamDelta(String)
    case thinkingDelta(String)
    case assistantMessage(content: String, thinking: String?)
    case toolCallStarted(name: String, input: JSONValue)
    case toolCallCompleted(name: String, result: AgentToolResult)
    case usageUpdate(inputTokens: Int, outputTokens: Int)
    case error(String)
    case done
}

File: android/pi-agent-core/src/main/kotlin/com/pimobile/agent/AgentTypes.kt
package com.pimobile.agent

import com.pimobile.ai.*
import kotlinx.serialization.json.JsonObject

interface Tool {
    val name: String
    val description: String
    val parametersSchema: JsonObject
    suspend fun execute(input: JsonObject): AgentToolResult
}

data class AgentToolResult(
    val toolCallId: String,
    val output: String,
    val details: ToolResultDetails? = null,
    val isError: Boolean = false
)

sealed class ToolResultDetails {
    data class File(val path: String, val content: String, val language: String? = null) : ToolResultDetails()
    data class Diff(val path: String, val hunks: List<DiffHunk>) : ToolResultDetails()
    data class Table(val columns: List<String>, val rows: List<List<String>>) : ToolResultDetails()
    data class Error(val message: String, val code: String? = null) : ToolResultDetails()
}

data class DiffHunk(
    val startLineOld: Int,
    val countOld: Int,
    val startLineNew: Int,
    val countNew: Int,
    val lines: List<DiffLine>
)

data class DiffLine(val type: DiffLineType, val content: String)

enum class DiffLineType { CONTEXT, ADD, REMOVE }

sealed class AgentEvent {
    data class StreamDelta(val delta: String) : AgentEvent()
    data class ThinkingDelta(val delta: String) : AgentEvent()
    data class AssistantMessage(val content: String, val thinking: String? = null) : AgentEvent()
    data class ToolCallStarted(val name: String, val input: JsonObject) : AgentEvent()
    data class ToolCallCompleted(val name: String, val result: AgentToolResult) : AgentEvent()
    data class UsageUpdate(val inputTokens: Int, val outputTokens: Int) : AgentEvent()
    data class Error(val message: String) : AgentEvent()
    data object Done : AgentEvent()
}
